ilastik/workflows/carving/opPreprocessing.py

```diff
@@ -413,18 +413,11 @@ def execute(self, slot, subindex, roi, result):
         self._dirty = False
         self.enableDownstream(True)
 
-        # copy over saved objects
-        if self.cachedResult[0] is not None:
-            mst.object_lut = self.cachedResult[0].object_lut
-            mst.object_names = self.cachedResult[0].object_names
-            mst.object_seeds_bg_voxels = self.cachedResult[0].object_seeds_bg_voxels
-            mst.object_seeds_fg_voxels = self.cachedResult[0].object_seeds_fg_voxels
-            mst.bg_priority = self.cachedResult[0].bg_priority
-            mst.no_bias_below = self.cachedResult[0].no_bias_below
-
         self.cachedResult = result
 
         result[0] = mst
+
+        self.PreprocessedData.setDirty(slice(None))
```

Comment:
not sure this is necessary, or useful - since we are already in `execute` and given the assert in the beginning of this function, we are requesting `self.PreprocessedData`. So in principle we would _not_ want to send any dirtyness upstream here (we just gave the data, setting the slot dirty invalidates it).

Reply:
PreprocessedData is the output of opPreprocessing, and feeds into opCarving's input `MST`.

I don't see any other way for opCarving to know that opPreprocessing.execute generated a new MST object. The question then is who tells opCarving that there is a new MST, and when:

1. When one of opPreprocessing's inputs change (in `opPreprocessing.propagateDirty`). This happens either when the Run button is pressed, or when we load a project file. Problem is that changed input implies a new MST needs to be generated, so this is why previously loading a project also caused it to re-run.
2. When the Run button is pressed. The preprocessingGui could give the request to opPreprocessing a callback to notify opCarving once it's done, or maybe the Workflow should be attaching this as a listener to the output slot?
3. Whenever a new MST is generated - which is what I did here

Reply:
> I don't see any other way for opCarving to know that opPreprocessing.execute generated a new MST object

when you are in execute, upstream already requested that data, so it "knew" it was dirty. So requesting the data from the slot is the reason we get into `execute` at all. Doing any dirty setting _after_ basically invalidates the result immediately - not sure why it even currently works.

Reply:
> we just gave the data

The caching implementation here basically runs against the intended use I think. execute hands over the currently cached data when `_dirty==False`, which is supposed to be the case whenever the downstream opCarving requests its MST input's value - i.e. the default use of execute().

It's only the special case when the Run button in preprocessingGui is clicked that opPreprocessing inputs change, hence `_dirty=True` and when the Gui itself makes a request for opPreprocessing's output.

Reply:
thx for the clarification (over zoom) @btbest :)